[DR-2785] Include DUOS firecloud group when retrieving a snapshot

okotsopoulos opened the pull request:
https://broadworkbench.atlassian.net/browse/DR-2785

When retrieving a snapshot, if it contains a pointer to a `duos_firecloud_group` entry, then return a representation of its associated DUOS Firecloud group.

<img width="1260" alt="Screen Shot 2022-10-25 at 4 30 32 PM" src="https://user-images.githubusercontent.com/79769153/197875775-9c536635-d691-400b-b67a-244f5c3affcf.png">

For now, the only impact on users is that the [retrieveSnapshot](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html#/snapshots/retrieveSnapshot) response body will include new field `duosFirecloudGroup`, which will be null in all cases.  Outside of tests, there is not currently a way to set a snapshot's `duos_firecloud_group_id` (but this is coming soon!)

I elected to leave `SnapshotSummary` and `SnapshotSummaryModel` unchanged (returned in snapshot enumeration) -- when we have a better sense of what Data Catalog might need to support DUOS integration end-to-end, we can revisit.

I broke up my commits for (hopefully) easier review, if you would prefer to step through them.

okotsopoulos committed:
Add link to duos_firecloud_group to snapshot table

okotsopoulos committed:
Include DuosFirecloudGroupModel when fetching snapshots

okotsopoulos committed:
DuosDao expansion and refactor

- Removed insertAndRetrieveFirecloudGroup as to not obscure generated row ID: insertFirecloudGroup now returns generated row ID
- Added method to retrieve by row ID as well as by DUOS ID, and shared common query
- Added deleteFirecloudGroup for testing
- Expanded unit tests

okotsopoulos committed:
Return associated DuosFirecloudGroupModel when retrieving snapshot

Expanded unit tests accordingly.

okotsopoulos commented on src/main/java/bio/terra/service/duos/DuosDao.java:
```diff
@@ -39,7 +48,7 @@ String getTdrServiceAccountEmail() {
 
   @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
-  public DuosFirecloudGroupModel insertAndRetrieveFirecloudGroup(
+  public UUID insertFirecloudGroup(
```
I originally thought that we'd store the DUOS ID of a linked DUOS dataset in the `snapshot` table.  But since we are storing our `duos_firecloud_group` surrogate key instead (which is autogenerated), I changed row insertion to return this generated key so that we have access to it.  This will prove useful when I build the flight to link a snapshot to a DUOS dataset.

okotsopoulos commented on src/main/java/bio/terra/service/snapshot/SnapshotDao.java:
```diff
@@ -421,14 +432,19 @@ Snapshot retrieveWorker(String sql, MapSqlParameterSource params) {
             .getSnapshotStorageAccount(snapshot.getId())
             .ifPresent(snapshot::storageAccountResource);
+
+        // Retrieve the DUOS Firecloud group associated with the snapshot.
```
Calling out that with the way this is written, we will always include the linked DUOS Firecloud group (if one exists) in the `retrieveSnapshot` response.

But another option is to expand the `SnapshotIncludeRetrieveModel` to allow the user to include / exclude it in the response.  LMK if this is preferred.

otchet-broad commented:
Silly questions: why is the Snapshot object being decorated with this information?  How are the downstream consumers using it?

okotsopoulos committed:
Only expose DUOS Firecloud group ID when retrieving snapshot

This key is a UUID pointing to a record in duos_firecloud_group.
Snapshot readers can retrieve snapshots, but we likely want to enforce a check for greater permissions to fetch more info on a snapshot's DUOS link.

okotsopoulos committed:
Forgot to commit changes to SnapshotServiceTest

okotsopoulos commented:
I am going to close this PR in favor of pursuing a PR with a flight to set / unset a snapshot's DUOS ID.  With that concrete use case, we'll be better positioned to evaluate what should be exposed in the snapshot object.
